src/test/java/io/github/christophermanahan/carnitas/RequestParserTest.java

```diff
@@ -0,0 +1,156 @@
+          + Headers.CONTENT_LENGTH + body.length()
+          + Constants.BLANK_LINE
+          + body;
+        Receiver receiver = new TestReceiver(request);
+        Parser parser = new RequestParser(new TestRequestBuilder());
+
+        Optional<Request> parsed = parser.parse(receiver);
+
+        String expectedRequest = requestMethod + body;
+        String parsedRequest = parsed.isEmpty() ? null : parsed.get().body();
+        assertEquals(expectedRequest, parsedRequest);
+    }
+
+    @Test
+    void parsesPOSTRequestWithBody() {
+        String requestMethod = "POST";
+        String body = "name=<something>";
+        String request = requestMethod +  " /simple_get " + Constants.VERSION + Constants.CRLF
+          + Headers.CONTENT_LENGTH + body.length()
+          + Constants.BLANK_LINE
+          + body;
+        Receiver receiver = new TestReceiver(request);
+        Parser parser = new RequestParser(new TestRequestBuilder());
+
+        Optional<Request> parsed = parser.parse(receiver);
+
+        String expectedRequest = requestMethod + body;
+        String parsedRequest = parsed.isEmpty() ? null : parsed.get().body();
+        assertEquals(expectedRequest, parsedRequest);
+    }
+
+    @Test
+    void parseIsEmptyIfReceiveFails() {
+        class FailedReceiver implements Receiver {
+            public String receiveLine() {
+                throw new RuntimeException();
+            }
+
+            public String receiveCharacters(int amount) {
+                return null;
+            }
+        }
+        Receiver receiver = new FailedReceiver();
+        Parser parser = new RequestParser(new TestRequestBuilder());
+
+        Optional<Request> parsed = parser.parse(receiver);
+
+        assertEquals(Optional.empty(), parsed);
+    }
+
+    private class TestReceiver implements Receiver {
+
+        private final Iterator<String> received;
+
+        TestReceiver(String request) {
+            List<String> lines = new ArrayList<>(Arrays.asList(request.split(Constants.CRLF)));
+            lines.add("");
+            this.received = lines.iterator();
+        }
+
+        public String receiveLine() {
+            return received.next();
+        }
+
+        public String receiveCharacters(int amount) {
+            return received.next();
+        }
+    }
+
+    private class TestRequestBuilder implements RequestBuilder {
+        private String requestMethod;
+        private String body;
+
+        public RequestBuilder requestMethod(String requestMethod) {
+            this.requestMethod = requestMethod;
+            return this;
+        }
+
+        public RequestBuilder body(String body) {
+            this.body = body;
+            return this;
+        }
+
+        public Request build() {
+            return body == null ? new TestRequest(requestMethod) : new TestRequest(requestMethod + body);
+        }
+    }
+
+    private class TestRequest implements Request {
```

Comment:
`HTTPRequest` doesn't have any behavior, does it? If not, you could probably use the real thing instead of a `TestRequest`